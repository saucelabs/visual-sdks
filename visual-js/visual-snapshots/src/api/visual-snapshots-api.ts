import { DiffingMethod, VisualApi } from "@saucelabs/visual";

export interface CreateVisualSnapshotsParams {
  branch: string;
  buildName: string;
  defaultBranch: string;
  project: string;
  customId: string;
  buildId: string;
}

export class VisualSnapshotsApi {
  private api: VisualApi;

  constructor(api: VisualApi) {
    this.api = api;
  }

  public async generateAndSendPdfFilSnapshots(
    pdfFilePages: AsyncGenerator<Buffer>,
    params: CreateVisualSnapshotsParams,
  ) {
    const buildId = await this.createBuild(params);

    let pageNumber = 1;
    for await (const pdfPageImage of pdfFilePages) {
      await this.uploadImageAndCreateSnapshot(
        pdfPageImage,
        buildId,
        pageNumber,
      );
      pageNumber++;
    }

    await this.finishBuild(buildId);
  }

  private async createBuild(
    params: CreateVisualSnapshotsParams,
  ): Promise<string> {
    const build = await this.api.createBuild({
      name: params.buildName,
      branch: params.branch,
      defaultBranch: params.defaultBranch,
      project: params.project,
      customId: params.customId,
    });
    console.info(`Build ${build.id} created.`);
    return build.id;
  }

  private async uploadImageAndCreateSnapshot(
    snapshot: Buffer,
    buildId: string,
    snapshotId: number,
  ) {
    const uploadId = await this.api.uploadSnapshot({
      buildId,
      image: { data: snapshot },
    });

    console.info(`Uploaded image to build ${buildId}: upload id=${uploadId}.`);

    const snapshotName = `page-${snapshotId}`;
    const snapshotMetadata = {
      diffingMethod: DiffingMethod.Balanced,
      buildUuid: buildId,
      name: snapshotName,
    };

    await this.api.createSnapshot({
      ...snapshotMetadata,
      buildUuid: buildId,
      uploadUuid: uploadId,
    });

    console.info(`Created a snapshot ${snapshotName} for build ${buildId}.`);
  }

  private async finishBuild(buildId: string) {
    await this.api.finishBuild({
      uuid: buildId,
    });
    console.info(`Build ${buildId} finished.`);
  }
}
