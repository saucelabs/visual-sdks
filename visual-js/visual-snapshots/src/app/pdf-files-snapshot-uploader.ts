export interface UploadPdfSnapshotsParams {
  readonly buildId: string;
  readonly pdfFilePaths: string[];
  readonly suiteNameFormat: string;
  readonly testNameFormat: string;
  readonly snapshotNameFormat: string;
}

export interface PdfSnapshotUploader {
  uploadSnapshots(params: UploadPdfSnapshotsParams): Promise<void>;
}
