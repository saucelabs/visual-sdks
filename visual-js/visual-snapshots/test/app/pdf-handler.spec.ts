import { VisualSnapshotsApi } from "../../src/api/visual-snapshots-api.js";
import { PdfSnapshotUploader } from "../../src/app/pdf-files-snapshot-uploader.js";
import {
  PdfCommandHandler,
  PdfCommandParams,
} from "../../src/app/pdf-handler.js";
import { MockFileExtractor } from "../mock-file-extractor.js";
import { mockLogger } from "../mock-logger.js";
import { __dirname } from "../system-helpers.js";
import path from "path";

describe("pdf-handler", () => {
  const createBuildMock = jest.fn<
    ReturnType<VisualSnapshotsApi["createBuild"]>,
    Parameters<VisualSnapshotsApi["createBuild"]>
  >();
  const finishBuildMock = jest.fn<
    ReturnType<VisualSnapshotsApi["finishBuild"]>,
    Parameters<VisualSnapshotsApi["finishBuild"]>
  >();

  const visualSnapshotsApi = {
    createBuild: createBuildMock,
    finishBuild: finishBuildMock,
  } as never as VisualSnapshotsApi;

  const uploadSnapshotsMock = jest.fn<
    ReturnType<PdfSnapshotUploader["uploadSnapshots"]>,
    Parameters<PdfSnapshotUploader["uploadSnapshots"]>
  >();

  const pdfSnapshotUploaderMock = {
    uploadSnapshots: uploadSnapshotsMock,
  } as never as PdfSnapshotUploader;

  const suiteName = "suiteName";
  const testName = "testName";
  const snapshotName = "snapshotName";
  const buildId = "buildId";

  const fileExtractor = new MockFileExtractor();

  const { logger, logged, reset: resetLogger } = mockLogger();

  beforeEach(() => {
    jest.resetAllMocks();
    resetLogger();
  });

  describe("creating build", () => {
    it("should create a build when buildId is not passed", async () => {
      fileExtractor.setFilesToReturn([
        "/absolute/path/to/files/1.pdf",
        "/absolute/path/to/files/2.pdf",
      ]);

      const handler = new PdfCommandHandler(
        visualSnapshotsApi,
        pdfSnapshotUploaderMock,
        fileExtractor,
        logger
      );

      const params: PdfCommandParams = {
        suiteName,
        testName,
        snapshotName,
        concurrency: 1,
      };

      await handler.handle(
        [path.join(__dirname(import.meta), "../files/")],
        params
      );

      expect(createBuildMock).toHaveBeenCalledWith(params);
      expect(logged).toEqual([
        {
          level: 30,
          msg: "Found 2 PDF files to process: /absolute/path/to/files/1.pdf, /absolute/path/to/files/2.pdf",
        },
      ]);
    });

    it("should not create a build when buildId is passed", async () => {
      fileExtractor.setFilesToReturn(["/absolute/path/to/files/1.pdf"]);

      const handler = new PdfCommandHandler(
        visualSnapshotsApi,
        pdfSnapshotUploaderMock,
        fileExtractor,
        logger
      );

      const params: PdfCommandParams = {
        suiteName,
        testName,
        snapshotName,
        buildId,
        concurrency: 1,
      };

      await handler.handle(
        [path.join(__dirname(import.meta), "../files/1.pdf")],
        params
      );

      expect(createBuildMock).not.toHaveBeenCalled();
      expect(logged).toEqual([
        {
          level: 30,
          msg: "Found 1 PDF file to process: /absolute/path/to/files/1.pdf",
        },
      ]);
    });

    it("should not create a build when there are no files to process", async () => {
      fileExtractor.setFilesToReturn([]);

      const handler = new PdfCommandHandler(
        visualSnapshotsApi,
        pdfSnapshotUploaderMock,
        fileExtractor,
        logger
      );

      const params: PdfCommandParams = {
        suiteName,
        testName,
        snapshotName,
        buildId,
        concurrency: 1,
      };

      await handler.handle(
        [path.join(__dirname(import.meta), "../files/1.pdf")],
        params
      );

      expect(createBuildMock).not.toHaveBeenCalled();
      expect(logged).toEqual([
        { level: 40, msg: "No PDF files found to process." },
      ]);
    });
  });

  describe("uploading snapshots", () => {
    it("should call uploadSnapshots with created build ID", async () => {
      fileExtractor.setFilesToReturn(["/absolute/path/to/files/1.pdf"]);

      const handler = new PdfCommandHandler(
        visualSnapshotsApi,
        pdfSnapshotUploaderMock,
        fileExtractor,
        logger
      );

      const params: PdfCommandParams = {
        suiteName,
        testName,
        snapshotName,
        concurrency: 1,
      };

      createBuildMock.mockResolvedValue(buildId);

      await handler.handle(
        [path.join(__dirname(import.meta), "../files/1.pdf")],
        params
      );

      expect(uploadSnapshotsMock).toHaveBeenCalledWith(
        expect.objectContaining({
          buildId,
          suiteNameFormat: suiteName,
          testNameFormat: testName,
          snapshotNameFormat: snapshotName,
        })
      );
    });

    it("should call uploadSnapshots with provided build ID", async () => {
      fileExtractor.setFilesToReturn(["/absolute/path/to/files/1.pdf"]);

      const handler = new PdfCommandHandler(
        visualSnapshotsApi,
        pdfSnapshotUploaderMock,
        fileExtractor,
        logger
      );

      const params: PdfCommandParams = {
        buildId,
        suiteName,
        testName,
        snapshotName,
        concurrency: 1,
      };

      await handler.handle(
        [path.join(__dirname(import.meta), "../files/1.pdf")],
        params
      );

      expect(uploadSnapshotsMock).toHaveBeenCalledWith(
        expect.objectContaining({
          buildId,
          suiteNameFormat: suiteName,
          testNameFormat: testName,
          snapshotNameFormat: snapshotName,
        })
      );
    });
  });

  describe("finishing build", () => {
    it("should finish build when buildId is not passed, using created build ID", async () => {
      createBuildMock.mockResolvedValue(buildId);

      fileExtractor.setFilesToReturn(["/absolute/path/to/files/1.pdf"]);

      const handler = new PdfCommandHandler(
        visualSnapshotsApi,
        pdfSnapshotUploaderMock,
        fileExtractor,
        logger
      );

      const params: PdfCommandParams = {
        suiteName,
        testName,
        snapshotName,
        concurrency: 1,
      };

      await handler.handle(
        [path.join(__dirname(import.meta), "../files/1.pdf")],
        params
      );

      expect(finishBuildMock).toHaveBeenCalledWith({ buildId });
    });

    it("should not finish build when buildId is passed", async () => {
      fileExtractor.setFilesToReturn(["/absolute/path/to/files/1.pdf"]);

      const handler = new PdfCommandHandler(
        visualSnapshotsApi,
        pdfSnapshotUploaderMock,
        fileExtractor,
        logger
      );

      const params: PdfCommandParams = {
        concurrency: 1,
        suiteName,
        testName,
        snapshotName,
        buildId,
      };

      await handler.handle(
        [path.join(__dirname(import.meta), "../files/1.pdf")],
        params
      );

      expect(finishBuildMock).not.toHaveBeenCalled();
    });
  });
});
