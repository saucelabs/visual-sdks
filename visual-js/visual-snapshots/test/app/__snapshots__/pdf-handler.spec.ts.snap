// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pdf-handler creating build should create a build when buildId is not passed 1`] = `
[
  {
    "count": 2,
    "level": 30,
    "msg": "Found PDF files to process",
    "paths": [
      "/absolute/path/to/files/1.pdf",
      "/absolute/path/to/files/2.pdf",
    ],
  },
]
`;

exports[`pdf-handler creating build should not create a build when buildId is passed 1`] = `
[
  {
    "count": 1,
    "level": 30,
    "msg": "Found PDF file to process",
    "paths": [
      "/absolute/path/to/files/1.pdf",
    ],
  },
]
`;

exports[`pdf-handler creating build should not create a build when there are no files to process 1`] = `
[
  {
    "level": 40,
    "msg": "No PDF files found to process.",
  },
]
`;
